hierarchicalforecast/core.py

```diff
@@ -90,170 +113,289 @@ class HierarchicalReconciliation:
-                     intervals_method: str = 'normality',
-                     sort_df: bool = True):
+
+    def _prepare_fit(
+        self,
+        Y_hat_nw: Frame,
+        S_nw: Frame,
+        Y_nw: Optional[Frame],
+        tags: Dict[str, np.ndarray],
+        level: Optional[List[int]] = None,
+        intervals_method: str = "normality",
+        id_col: str = "unique_id",
+        time_col: str = "ds",
+        target_col: str = "y",
+    ):
         """
         Performs preliminary wrangling and protections
         """
-        #-------------------------------- Match Y_hat/Y/S index order --------------------------------#
-        if sort_df:
-            Y_hat_df = Y_hat_df.reset_index()
-            Y_hat_df.unique_id = Y_hat_df.unique_id.astype('category')
-            Y_hat_df.unique_id = Y_hat_df.unique_id.cat.set_categories(S_df.index)
-            Y_hat_df = Y_hat_df.sort_values(by=['unique_id', 'ds'])
-            Y_hat_df = Y_hat_df.set_index('unique_id')
-
-            if Y_df is not None:
-                Y_df = Y_df.reset_index()
-                Y_df.unique_id = Y_df.unique_id.astype('category')
-                Y_df.unique_id = Y_df.unique_id.cat.set_categories(S_df.index)
-                Y_df = Y_df.sort_values(by=['unique_id', 'ds'])
-                Y_df = Y_df.set_index('unique_id')
-
-            S_df.index = pd.CategoricalIndex(S_df.index, categories=S_df.index)
-
-        #----------------------------------- Check Input's Validity ----------------------------------#
+        Y_hat_nw_cols = Y_hat_nw.columns
+        S_nw_cols = S_nw.columns
+
+        # -------------------------------- Match Y_hat/Y/S index order --------------------------------#
+        # TODO: This is now a bit slow as we always sort.
+        S_nw = S_nw.with_columns(**{f"{id_col}_id": np.arange(len(S_nw))})
+
+        Y_hat_nw = Y_hat_nw.join(S_nw[[id_col, f"{id_col}_id"]], on=id_col, how="left")
+        Y_hat_nw = Y_hat_nw.sort(by=[f"{id_col}_id", time_col])
+        Y_hat_nw = Y_hat_nw[Y_hat_nw_cols]
+
+        if Y_nw is not None:
+            Y_nw_cols = Y_nw.columns
+            Y_nw = Y_nw.join(S_nw[[id_col, f"{id_col}_id"]], on=id_col, how="left")
+            Y_nw = Y_nw.sort(by=[f"{id_col}_id", time_col])
+            Y_nw = Y_nw[Y_nw_cols]
+
+        S_nw = S_nw[S_nw_cols]
+
+        # ----------------------------------- Check Input's Validity ----------------------------------#
+
         # Check input's validity
-        if intervals_method not in ['normality', 'bootstrap', 'permbu']:
-            raise ValueError(f'Unkwon interval method: {intervals_method}')
+        if intervals_method not in ["normality", "bootstrap", "permbu"]:
+            raise ValueError(f"Unknown interval method: {intervals_method}")
+
+        if self.insample or (intervals_method in ["bootstrap", "permbu"]):
+            if Y_nw is None:
+                raise Exception("You need to provide `Y_df`.")
```

Comment:
```suggestion
        if (self.insample or intervals_method in ["bootstrap", "permbu"]) and Y_df is None:
            raise ValueError("You need to provide `Y_df`.")
```

Reply:
NB: `self.insample` is whether any method is insample, but whether Y_df is _not_ None varies across all methods being reconciled, so making your suggested change will lead to failures. (e.g. in the case where you have MinT with insample and a BottomUp)

The behavior is confusing, for now I've added a TODO comment to make it method-specific rather than partially across all methods, partially for a single method.